parsing/builtin_attributes.ml

```diff
@@ -15,6 +15,72 @@
 
 open Asttypes
 open Parsetree
+open Ast_helper
+
+
+module Attribute_table = Hashtbl.Make (struct
+  type t = string with_loc
+
+  let hash : t -> int = Hashtbl.hash
+  let equal : t -> t -> bool = (=)
+end)
+let unused_attrs = Attribute_table.create 128
+let mark_used t = Attribute_table.remove unused_attrs t
+
+(* [attr_order] is used to issue unused attribute warnings in the order the
+   attributes occur in the file rather than the random order of the hash table
+*)
+let attr_order a1 a2 =
+  match String.compare a1.loc.loc_start.pos_fname a2.loc.loc_start.pos_fname
+  with
+  | 0 -> Int.compare a1.loc.loc_start.pos_cnum a2.loc.loc_start.pos_cnum
+  | n -> n
+
+let warn_unused () =
+  let keys = List.of_seq (Attribute_table.to_seq_keys unused_attrs) in
+  let keys = List.sort attr_order keys in
+  List.iter (fun sloc ->
+    Location.prerr_warning sloc.loc (Warnings.Misplaced_attribute sloc.txt))
+    keys
+
+(* These are the attributes that are tracked in the builtin_attrs table for
+   misplaced attribute warnings. *)
+let builtin_attrs =
+  [ "alert"; "ocaml.alert"
+  ; "boxed"; "ocaml.boxed"
+  ; "deprecated"; "ocaml.deprecated"
+  ; "deprecated_mutable"; "ocaml.deprecated_mutable"
+  ; "explicit_arity"; "ocaml.explicit_arity"
+  ; "immediate"; "ocaml.immediate"
+  ; "immediate64"; "ocaml.immediate64"
+  ; "inline"; "ocaml.inline"
+  ; "inlined"; "ocaml.inlined"
+  ; "noalloc"; "ocaml.noalloc"
+  ; "ppwarning"; "ocaml.ppwarning"
+  ; "tailcall"; "ocaml.tailcall"
+  ; "unboxed"; "ocaml.unboxed"
+  ; "untagged"; "ocaml.untagged"
+  ; "unrolled"; "ocaml.unrolled"
+  ; "warnerror"; "ocaml.warnerror"
+  ; "warning"; "ocaml.warning"
+  ; "warn_on_literal_pattern"; "ocaml.warn_on_literal_pattern"
+  ]
```

Comment:
I've made this change, and the similar suggested changes below.  This resulted in some additional simplifications (various functions that used to take lists can now just take a single attribute name).

I was on the fence, after reading the suggestion, because a consequence of this move is that we'll needlessly allocate additional strings when we strip `ocaml.` prefixes.  However, after seeing the simplifications, I'm happy with the change.  (Also, for small programs or programs with no attributes, this is probably just a win.)